File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        queue<int> q;
        unordered_map<char,int> m;
        int maxLen = 0;
        for(auto i: s) {
            if(m.find(i)!=m.end() && m[i]!=0) {
                int currLen = q.size();
                maxLen = max(maxLen, currLen);
                while(q.front()!=i) {
                    char ch = q.front();
                    m[ch]--;
                    q.pop();
                }
                q.pop();
                q.push(i);
            } else {
                m[i]++;
                q.push(i);
            }
        }
        int currLen = q.size();
        maxLen = max(maxLen, currLen);
        return maxLen;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int[] hash = new int[256];
        Arrays.fill(hash, -1);
        int maxLen = 0;
        int start = 0;
        for(int i=0; i<s.length(); i++) {
            char ch = s.charAt(i);
            if(hash[ch]!=-1) {
                start = Math.max(start, hash[ch] + 1);
            } 
            hash[ch] = i;
            maxLen = Math.max(maxLen, i-start+1);
        }

        return maxLen;
    }
}